tiledb/sm/filesystem/s3.cc

```diff
@@ -1507,12 +1496,12 @@ Status S3::init_client() const {
   static std::mutex static_client_init_mtx;
   {
     std::lock_guard<std::mutex> static_lck(static_client_init_mtx);
-    if (credentials_provider_ == nullptr) {
+    if (credentials_provider == nullptr) {
       client_ =
           make_shared<TileDBS3Client>(HERE(), s3_params_, *client_config_);
     } else {
       client_ = make_shared<TileDBS3Client>(
-          HERE(), s3_params_, credentials_provider_, *client_config_);
+          HERE(), s3_params_, credentials_provider, *client_config_);
```

Comment:
Can you please double check that `Aws::S3::S3Client` doesn't hold a raw ptr or some sort of weak ptr for `credentials_provider` so we are sure the lifetime of your new local variable is extended properly?

Reply:
[Through](https://github.com/aws/aws-sdk-cpp/blob/f187cf68c15062d88bd6b21c56a92a19bf9db6c9/generated/src/aws-cpp-sdk-s3/source/S3Client.cpp#L253) [several](https://github.com/aws/aws-sdk-cpp/blob/f187cf68c15062d88bd6b21c56a92a19bf9db6c9/generated/src/aws-cpp-sdk-s3/source/S3ExpressSignerProvider.cpp#L19) [layers](https://github.com/aws/aws-sdk-cpp/blob/f187cf68c15062d88bd6b21c56a92a19bf9db6c9/src/aws-cpp-sdk-core/source/auth/signer-provider/DefaultAuthSignerProvider.cpp#L27) of [indirection](https://github.com/aws/aws-sdk-cpp/blob/f187cf68c15062d88bd6b21c56a92a19bf9db6c9/src/aws-cpp-sdk-core/source/auth/signer/AWSAuthV4Signer.cpp#L61), the credentials provider is [stored](https://github.com/aws/aws-sdk-cpp/blob/f187cf68c15062d88bd6b21c56a92a19bf9db6c9/src/aws-cpp-sdk-core/include/aws/core/auth/signer/AWSAuthV4Signer.h#L209) in a shared pointer.